pkg/dependenciesdistributor/dependencies_distributor.go

```diff
@@ -676,7 +681,8 @@ func buildAttachedBinding(binding *workv1alpha2.ResourceBinding, object *unstruc
 				Name:            object.GetName(),
 				ResourceVersion: object.GetResourceVersion(),
 			},
-			RequiredBy: result,
+			RequiredBy:         result,
+			ConflictResolution: binding.Spec.ConflictResolution,
```

Comment:
Seems the `attachedBinding` could be shared by multiple parent ResourceBindings, am I right?
What if the `conflict resolution` is different between the parent RBs?

Reply:
Good question!

Reply:
Currently, there is no solution available.

Reply:
``` diff
--- a/pkg/apis/work/v1alpha2/binding_types.go
+++ b/pkg/apis/work/v1alpha2/binding_types.go
@@ -203,6 +203,20 @@ type TargetCluster struct {
        Replicas int32 `json:"replicas,omitempty"`
 }
 
+// RequiredByTargetCluster represents the identifier of a member cluster that is required by the resource.
+type RequiredByTargetCluster struct {
+       // Name of target cluster.
+       Name string `json:"name"`
+       // Replicas in target cluster
+       // +optional
+       Replicas int32 `json:"replicas,omitempty"`
+
+       // ConflictResolution declares how potential conflict should be handled when
+       // If ConflictResolution is nil, it means it is consistent with the ResourceBinding's ConflictResolution.
+       // +optional
+       ConflictResolution *policyv1alpha1.ConflictResolution `json:"conflictResolution,omitempty"`
+}
+
 // GracefulEvictionTask represents a graceful eviction task.
 type GracefulEvictionTask struct {
        // FromCluster which cluster the eviction perform from.
@@ -273,7 +287,7 @@ type BindingSnapshot struct {
 
        // Clusters represents the scheduled result.
        // +optional
-       Clusters []TargetCluster `json:"clusters,omitempty"`
+       Clusters []RequiredByTargetCluster `json:"clusters,omitempty"`
 }


```



t seems that this is the only way to modify it.

@RainbowMango @XiShanYongYe-Chang @jwcesign WDYT.

Reply:
or

``` diff
diff --git a/pkg/apis/work/v1alpha2/binding_types.go b/pkg/apis/work/v1alpha2/binding_types.go
index 554218f11..9cb5bf5a3 100644
--- a/pkg/apis/work/v1alpha2/binding_types.go
+++ b/pkg/apis/work/v1alpha2/binding_types.go
@@ -201,6 +201,11 @@ type TargetCluster struct {
        // Replicas in target cluster
        // +optional
        Replicas int32 `json:"replicas,omitempty"`
+
+       // ConflictResolution declares how potential conflict should be handled when
+       // If ConflictResolution is nil, it means it is consistent with the ResourceBinding's ConflictResolution.
+       // +optional
+       ConflictResolution *policyv1alpha1.ConflictResolution `json:"conflictResolution,omitempty"`
 }
 



```

Reply:
The essence of this issue is that we assume a resource uses the same `conflict resolution` in all member clusters by default. However, a better approach would be to adopt different `conflict resolution` for each member cluster.

Reply:
It seems that even if the conflict strategy is refined to the granularity of clusters, inconsistent situations will still occur if dependent resources with different conflict strategies are propagated to the same cluster. Overall, the size of the conflict domain has decreased, but the overall complexity of the solution has increased.

If there are multiple dependent resources with different conflict strategies, a more assertive choice is to set the conflict strategy of the current resource as `Overwrite`; or a more relaxed option is to set it as `Abort`. Wait for more user feedback in the future before considering whether to refine it at the cluster level.

Reply:
> It seems that even if the conflict strategy is refined to the granularity of clusters, inconsistent situations will still occur if dependent resources with different conflict strategies are propagated to the same cluster

No, if the same resource in the same cluster has different conflict strategies, that is, one is `Abort` and the other is `Overwrite`, then the default behavior is actually `Overwrite`.

Reply:
How to explain this behavior? If we follow this default method, can it provide us with a basis for explaining the conflict strategy of overall resources?

Reply:
>.  If we follow this default method, can it provide us with a basis for explaining the conflict strategy of overall resources?

I don't understand your meaning.

Reply:
> No, if the same resource in the same cluster has different conflict strategies, that is, one is Abort and the other is Overwrite, then the default behavior is actually Overwrite.



This is an obvious behavior. Once there is an `Overwrite`, the resources of the control plane should overwrite the resources of the member clusters.

Reply:
Sorry, I didn't express myself clearly. Let me try to explain my point again. My idea is whether we can apply the rules behind this behavior to the selection of conflict strategies for resources as well. For example, if there are more than two dependent resources with different conflict strategies, then we will consider the conflict strategy of the current resource as Overwrite or Abort.

Reply:
The root cause is that the dependent resources(like ConfigMap) support reuse by multiple parent resources(like Deployments), that inconsistent with the constraints of PropagationPolicy. I think we need to review the original design first, to see if we need to change the behavior. @chaunceyjiang Do you have any concerns about the behavior change?

Reply:
>  I think we need to review the original design first, to see if we need to change the behavior.

+1 


> @chaunceyjiang Do you have any concerns about the behavior change?


Currently there is none.

Reply:
Ok